yourbench/pipeline/chunking.py

```diff
@@ -0,0 +1,608 @@
+import random
+import re
+from typing import Any, Dict, List, Optional
+
+import evaluate
+import matplotlib.pyplot as plt
+import torch
+import torch.nn.functional as F
+from loguru import logger
+from transformers import AutoModel, AutoTokenizer
+
+from yourbench.utils.dataset_engine import (
+    save_dataset,
+    smart_get_output_dataset_name,
+    smart_get_output_subset,
+    smart_get_source_dataset_name,
+    smart_get_source_subset,
+    smart_load_dataset,
+)
+
+
+# === Stage-Specific Logger Configuration ===
+os.makedirs("logs", exist_ok=True)
+logger.add("logs/chunking.log", level="DEBUG", rotation="10 MB", enqueue=True, backtrace=True, diagnose=True)
+
+try:
+    # Attempt to load perplexity metric from evaluate
+    _perplexity_metric = evaluate.load("perplexity", module_type="metric", model_id="gpt2")
+    logger.info("Loaded 'perplexity' metric with model_id='gpt2'.")
+except Exception as perplexity_load_error:
+    logger.warning(
+        "Could not load perplexity metric from 'evaluate'. Skipping perplexity. Error: {}", perplexity_load_error
+    )
+    _perplexity_metric = None
+
+try:
+    # Attempt to import textstat for readability metrics
+    import textstat
+
+    _use_textstat = True
+except ImportError:
+    logger.warning("Package 'textstat' not installed. Readability metrics will be skipped.")
+    _use_textstat = False
+
+# Ensure local plots directory
+os.makedirs("plots", exist_ok=True)
+
+
+def run(config: Dict[str, Any]) -> None:
+    """
+    Main pipeline entry point for the chunking stage.
+
+    Args:
+        config (Dict[str, Any]): The entire pipeline configuration dictionary.
+
+    Returns:
+        None. This function saves the updated dataset containing chunked
+        documents to disk or the Hugging Face Hub, based on the config.
+
+    Raises:
+        RuntimeError: If a critical error is encountered that prevents chunking.
+                      The error is logged, and execution attempts a graceful exit.
+    """
+    # Retrieve chunking configuration from config
+    chunking_config = config.get("pipeline", {}).get("chunking", {})
+    if not chunking_config.get("run", False):
+        logger.info("Chunking stage is disabled. Skipping.")
+        return
+
+    logger.info("Starting chunking stage...")
+
+    # Attempt to load dataset
+    try:
+        source_dataset_name = smart_get_source_dataset_name("chunking", config)
+        source_subset = smart_get_source_subset("chunking", config)
+        output_dataset_name = smart_get_output_dataset_name("chunking", config)
+        output_subset = smart_get_output_subset("chunking", config)
+
+        dataset = smart_load_dataset(source_dataset_name, config, source_subset)
+        logger.debug("Loaded dataset '{}' with {} rows for chunking.", source_dataset_name, len(dataset))
+    except Exception as ds_error:
+        logger.error("Failed to load dataset: {}", ds_error)
+        logger.warning("Chunking stage cannot proceed. Exiting.")
```

Comment:
If a step completely fails, we want it to stop the whole process since each step is based on the previous one - we also want to be able to launch with a cached dataset

Reply:
i added an error here